src/sentry/tasks/integrations/migrate_alert_rules.py

```
@@ -0,0 +1,109 @@
+from sentry.tasks.integrations import logger
+
+ALERT_LEGACY_INTEGRATIONS = {
+    "id": "sentry.rules.actions.notify_event.NotifyEventAction",
+    "name": "Send a notification (for all legacy integrations)",
+}
+
+
+@instrumented_task(
+    name="sentry.tasks.integrations.migrate_alert_rules",
+    queue="integrations",
+    default_retry_delay=60 * 5,
+    max_retries=5,
+)
+@retry(exclude=[Integration.DoesNotExist])
+def migrate_alert_rules(integration_id: int, organization_id: int) -> None:
+    from sentry_plugins.opsgenie.plugin import OpsGeniePlugin
+
+    integration, organization_integration = integration_service.get_organization_context(
+        organization_id=organization_id, integration_id=integration_id
+    )
+    if not integration:
+        raise Integration.DoesNotExist
+    if not organization_integration:
+        raise OrganizationIntegration.DoesNotExist
+
+    seen_keys = {team["integration_key"] for team in organization_integration.config["team_table"]}
+
+    for project in Project.objects.filter(organization_id=organization_id):
+        plugin = None
+        for p in plugins.for_project(project):
+            if isinstance(p, OpsGeniePlugin) and p.is_configured(project):
+                plugin = p
+                break
+        if not plugin:
+            continue
+
+        # migrate key
+        api_key = plugin.get_option("api_key", project)
+        if api_key not in seen_keys:
+            seen_keys.add(api_key)
+            config = organization_integration.config
+            team_table = config["team_table"]
+            team = {
+                "team": project.name + " [MIGRATED]",
+                "id": str(organization_id) + "-" + project.name,
+                "integration_key": api_key,
+            }
+            team_table.append(team)
+            config.update({"team_table": team_table})
+            with transaction.atomic(router.db_for_write(OrganizationIntegration)):
```

So looking at this again, we won't be able to use transactions here. The two database edits for adding a new key and updating alert rules occur in two different silos (for hybrid cloud purposes) so there is no way to wrap them in `transaction.atomic`.

Instead, we should raise an error if the RPC call to `update_organization_integration` fails, that way the task is retried.

---

This should address the cross-silo concern @dashed  pointed out as well :)